src/test/java/com/google/devtools/build/lib/server/GrpcServerTest.java

```diff
@@ -460,6 +469,320 @@ public void onCompleted() {
+              List<String> args,
+              OutErr outErr,
+              LockingMode lockingMode,
+              String clientDescription,
+              long firstContactTimeMillis,
+              Optional<List<Pair<String, String>>> startupOptionsTaggedWithBazelRc)
+              throws InterruptedException {
+            if (args.contains(firstCommandArg)) {
+              OutputStream out = outErr.getOutputStream();
+              try {
+                while (barsReceived.get() < 3) {
+                  if (Thread.interrupted()) {
+                    return BlazeCommandResult.failureDetail(
+                        FailureDetail.newBuilder()
+                            .setInterrupted(
+                                Interrupted.newBuilder().setCode(Code.INTERRUPTED))
+                            .build());
+                  }
+                  out.write(new byte[1024]);
+                }
+                return BlazeCommandResult.success();
+              } catch (IOException e) {
+                throw new IllegalStateException(e);
+              }
+            } else {
+              OutputStream out = outErr.getOutputStream();
+              try {
+                while (barsReceived.get() < 3) {
+                  if (Thread.interrupted()) {
+                    return BlazeCommandResult.failureDetail(
+                        FailureDetail.newBuilder()
+                            .setInterrupted(
+                                Interrupted.newBuilder().setCode(Code.INTERRUPTED))
+                            .build());
+                  }
+                  out.write(new byte[1024]);
+                }
+                return BlazeCommandResult.success();
+              } catch (IOException e) {
+                throw new IllegalStateException(e);
+              }
+            }
+          }
+        };
+    createServer(dispatcher);
+
+    CountDownLatch gotFoo = new CountDownLatch(1);
+    CountDownLatch finishedBar = new CountDownLatch(1);
+    AtomicReference<RunResponse> lastFooResponse = new AtomicReference<>();
+    AtomicReference<RunResponse> lastBarResponse = new AtomicReference<>();
+
+    CommandServerStub stub = CommandServerGrpc.newStub(channel);
+    stub.run(
+        createRequest(firstCommandArg),
+        new StreamObserver<RunResponse>() {
+          @Override
+          public void onNext(RunResponse value) {
+            gotFoo.countDown();
+            lastFooResponse.set(value);
+          }
+
+          @Override
+          public void onError(Throwable t) {}
+
+          @Override
+          public void onCompleted() {}
+        });
+
+    // Wait for the first command to startup
+    gotFoo.await();
+
+    stub.run(
+        createRequest(secondCommandArg),
+        new StreamObserver<RunResponse>() {
+          @Override
+          public void onNext(RunResponse value) {
+            if (value.getFinished()) {
+              finishedBar.countDown();
```

Comment:
should this go in onCompleted or onError? (see previous comment about this)

Reply:
Removed now.